arch-power: Add Initial Support for VSX Instructions (mtvsrd, mfvsrd) in Power ISA

divyangt-07 opened the pull request:
Summary

This Pull Request introduces initial support for the `mtvsrd` and `mfvsrd` instructions within the Power ISA in gem5. These instructions belong to the VSX (Vector-Scalar Extension) instruction set and represent a step toward broader VSX coverage.
-----------------------

Implementation Details

a. Implemented decode logic and execution behavior for `mtvsrd` and `mfvsrd`
b. Updated instruction formatting and register handling in relevant ISA files
c. Integrated bitfield mapping and symbol resolution
d. Temporarily disabled vector debug flags due to unresolved initialization issues
------------

Context

This work stems from discussions during Gem5 BootCamp 2024 and aims to contribute to ongoing support for the Power ISA's VSX feature set. Further instructions and full vector support are planned.
----------

Request for Review

@power-isa-maintainers, I am seeking your feedback on the following:
a. Symbol resolution conventions used for VSX
b. Register file usage (scalar vs. vector overlap handling)
c. Whether the approach aligns with existing ISA extension patterns

Please let me know if any structural changes are required or if there's an internal checklist you would prefer me to follow.

Thank you!
Divyang

divyangt-07 committed:
arch-power: Add mtvsrd and mfvsrd instructions

Implemented the mtvsrd and mfvsrd instructions in decoder.isa.
Added their corresponding C++ functions in vector_scalar.cc.
Fixed related formatting issues flagged by the style checker.

divyangt-07 committed:
arch-power: Grouped the mtvsrd and mfvsrd instructions

The mtvsrd and mfvsrd instructions are grouped under the common opcode

divyangt-07 committed:
arch-power: Updated the SConscript

divyangt-07 committed:
arch-power: Added the .cc and .hh file for Vector Scalar Insts

The mtvsrd and mfvsrd instructions are implemented. The vector_scalar.hh file includes the function definitons and vector_scalar.cc file includes the C++ functions of the respective instructions

divyangt-07 committed:
arch-power: VSX Instructions

Updated the mtvsrd and mfvsrd C++ functions to resolve the compilation errors.

divyangt-07 committed:
arch-power: VSX Instructions

Updated the mtvsrd and mfvsrd C++ functions to resolve the compilation errors.

divyangt-07 committed:
arch-power: VSX Instructions

Updated the mtvsrd and mfvsrd C++ functions to resolve the compilation errors.

divyangt-07 committed:
arch-power: VSX Instructions

Updated the bitfields for the mtvsrd and mfvsrd instructions.

divyangt-07 committed:
arch-power: VSX Instructions

Updated the decoder section for the mtvsrd and mfvsrd instructions to resolve the compilation errors.

divyangt-07 committed:
arch-power: VSX Instructions

Added the vector_scalar.isa in the include section

divyangt-07 committed:
arch-power: VSX Instructions

Updated the file to resolve the compilation errors.

divyangt-07 committed:
arch-power: VSX Instructions

Added the vector_scalar.isa file to take into account the VSX instrcutions.

divyangt-07 committed:
arch-power: VSX Instructions

Added the vsx files file to take into account the VSX instrcutions.

divyangt-07 committed:
arch-power: VSX Instructions

Added the vector register operands to take into account the VSX instrcutions.

divyangt-07 committed:
arch-power: VSX Instructions

Created a vec.hh file by referencing the int.hh existing for the Arithmetic Instructions.

powerjg commented on src/arch/power/regs/vec.hh:
```diff
@@ -0,0 +1,113 @@
+#ifndef __ARCH_POWER_REGS_VEC_HH__
```
Needs copyright

divyangt-07 replied:
Added the appropriate Copyright to all the new files.

powerjg replied:
Use /* */

powerjg commented on src/arch/power/isa/formats/vector_scalar.isa:
```diff
@@ -0,0 +1,4 @@
+def format VSRTransferOp(code, inst_flags = []) {{
```
needs copyright

divyangt-07 replied:
Added the appropriate Copyright to all the new files.

powerjg commented on src/arch/power/insts/vector_scalar.hh:
```diff
@@ -0,0 +1,20 @@
+#ifndef __ARCH_POWER_INSTS_VECTOR_SCALAR_HH__
```
needs copyright

divyangt-07 replied:
Added the appropriate Copyright to all the new files.

powerjg replied:
In c++ files the correct comment for the copyright is /* */. Use the same style as the .cc file above.

powerjg commented on src/arch/power/insts/vector_scalar.cc:
```diff
@@ -0,0 +1,56 @@
+#include "arch/power/insts/misc.hh"
```
needs copyright

divyangt-07 replied:
Added the appropriate Copyright to all the new files.

powerjg commented on src/arch/power/insts/vector_scalar.cc:
```diff
@@ -0,0 +1,56 @@
+#include "cpu/thread_context.hh"
+#include "vector_scalar.hh"
+
+using namespace gem5;
```
rather than `using`, add `namespace gem5 {`

powerjg commented on src/arch/power/insts/vector_scalar.cc:
```diff
@@ -0,0 +1,56 @@
+typedef __uint128_t uint128_t;
+
+// Register classes are in PowerISA namespace (from src/arch/power/regs/)
+using namespace PowerISA;
```
ditto above

powerjg commented on src/arch/power/insts/vector_scalar.cc:
```diff
@@ -0,0 +1,56 @@
+using namespace gem5;
+
+// Define 128-bit unsigned int if not already available
+typedef __uint128_t uint128_t;
```
is this compatible with all compilers?

divyangt-07 replied:
In my opinion, it is not.
Hence, updated to add a layer to check:
// Define 128-bit unsigned int if not already available
#if defined(__SIZEOF_INT128__)
typedef __uint128_t uint128_t;
#else
#warning "128-bit integer type not supported on this compiler"
#endif

powerjg commented on src/arch/power/insts/vector_scalar.cc:
```diff
@@ -0,0 +1,56 @@
+// Register classes are in PowerISA namespace (from src/arch/power/regs/)
+using namespace PowerISA;
+
+void mtvsrd_exec(ThreadContext *tc, int t_s, int ra_vsx, int tx_sx) {
```
style is to put the return type on the prior line and the curly braces on its own line.

divyangt-07 replied:
Updated the function definition to match the gem5 style.

powerjg reviewed (commented):
FYI, there are no POWER maintainers right now. 

Other than missing copyright and some style issues, the code looks fine to me. My only big question is how was this tested?

divyangt-07 commented:
> FYI, there are no POWER maintainers right now.
> 
> Other than missing copyright and some style issues, the code looks fine to me. My only big question is how was this tested?

divyangt-07 committed:
arch-power: VSX Instructions

Updated the Copyright

divyangt-07 committed:
arch-power: VSX Instructions

Updated the Copyright

divyangt-07 committed:
arch-power: VSX Instructions

Updated the Copyright

divyangt-07 committed:
arch-power: VSX Instructions

Updated the Copyright

divyangt-07 commented:
> FYI, there are no POWER maintainers right now.
> 
> Other than missing copyright and some style issues, the code looks fine to me. My only big question is how was this tested?

Thank you for the feedback!

The vsx_instruction implementation was done by taking the existing arithmetic instruction implementations in the POWER ISA as reference. At this point, I’m facing some compilation issues related to the decoder.cc file generation. I created this Pull Request primarily to get some guidance on how best to move forward, since I ran into blockers during the build process.

I had also created a discussion post back in April, but unfortunately it didn’t gain much traction. Given the absence of active POWER maintainers, I thought this PR might be a more effective way to reach out to others in the community who might be able to help.

Currently, I’m working on resolving the compilation issues to get a successful build, and I plan to continue iterating on the changes based on feedback and what’s needed to complete the implementation. I’ll keep committing updates as I make progress.

Thanks again!

divyangt-07 committed:
arch-power: VSX Instructions

Uploading a README file and compilation error signature file as of June 22, 2025

divyangt-07 committed:
arch-power: VSX Instructions

Changed the comment type to ensure no compilation errors.

divyangt-07 committed:
arch-power: VSX Instructions

Updated the vector_scalar.cc file to incorporate the feedback provided in the PR request.

divyangt-07 committed:
arch-power: VSX Instructions

Updated the generation of the C++ files to resolve the compilation errors in building the Gem5 for Power ISA. With this update the gem5 build is successful.

divyangt-07 committed:
arch-power: VSX Instructions

Unit Level Testing of the 'mtvsrd' and 'mfvsrd' instructions.

divyangt-07 committed:
arch-power: VSX Instructions

READEME_VSX_INSTS file updated. Added the Revision History to it.

divyangt-07 committed:
arch-power: VSX Instructions

READEME_VSX_INSTS file updated the Notes section to add the run-time error faced details.

divyangt-07 committed:
arch-power: VSX Instructions

Craeted a .txt file that keeps the track of the gem5 run-time errors.

powerjg commented on tests/power_test/mtvsrd_mfvsrd_test.c:
```diff
@@ -0,0 +1,29 @@
+#include <inttypes.h>
```
Needs copyright.

erin-le commented:
@divyangt-07: Hello, thanks for your contribution! Could you change this PR's target branch to `develop`? You should be able to change it by hitting the `edit` button by the title of the PR:
![Screenshot 2025-07-03 at 5 48 10 PM](https://github.com/user-attachments/assets/eee915be-fbbd-4aa4-8f3b-b0c2fd019057)

Could you also add `arch-power: ` to the start of the PR's title? Thanks!

erin-le commented:
@divyangt-07: Hello, could you respond to the above comments when you get a chance? Thank you!

divyangt-07 commented:
> @divyangt-07: Hello, thanks for your contribution! Could you change this PR's target branch to `develop`? You should be able to change it by hitting the `edit` button by the title of the PR: ![Screenshot 2025-07-03 at 5 48 10 PM](https://private-user-images.githubusercontent.com/122753811/462254076-eee915be-fbbd-4aa4-8f3b-b0c2fd019057.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTMzMzMzNTcsIm5iZiI6MTc1MzMzMzA1NywicGF0aCI6Ii8xMjI3NTM4MTEvNDYyMjU0MDc2LWVlZTkxNWJlLWZiYmQtNGFhNC04ZjNiLWIwYzJmZDAxOTA1Ny5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcyNFQwNDU3MzdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jMDJkMWQyYmU0ZDNmMTJhNTY1MDg4YWNlYzNmNjBmN2RiMGVlMWYwNTFhZjZlMWE4M2RjNTA2Y2Q0MTRmZjU1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.ZMQj3YR6NjgidKa20X0mDSUUCqGqAvTz9RDYJrrFpIM)
> 
> Could you also add `arch-power: ` to the start of the PR's title? Thanks!

Hi,
Yes, I have changed the title of the Pull Request and also updated the PR's target branch to develop.

erin-le commented:
@divyangt-07: Hello, in addition to what you've done so far, can you:
1. Address powerjg's comments
2. Resolve the conflicts between this PR and `develop` and ensure the CI tests pass afterward
3. Remove the files `build_errors_June22.txt`, `gem5_run_output_June29.txt`, and `src/arch/README_VSX_INSTS`

Thank you!

erin-le commented:
@divyangt-07: Hello, just another reminder about this PR. Could you address the concerns in the previous comment when you get a chance? Thank you!

erin-le commented:
@divyangt-07: Hello, just another reminder to address the comments on this PR and make the requested changes. Thank you!

erin-le commented:
@divyangt-07: Hello, thanks for this contribution. Would you be okay with closing this PR for now, if you don't have the time to work on it, and reopening it later?

erin-le commented:
This PR is being closed for inactivity. If you would like to continue working on it in the future, please feel free to reopen this PR. Thank you!
